keps/sig-apps/3998-job-success-completion-policy/README.md

```diff
@@ -0,0 +1,913 @@
+	// for the job. When SucceededCount is used along with SucceededIndexes, the check is 
+	// constrained only to the set of indexes specified by SucceededIndexes.
+	// When this field is null, this field doesn't default to any value
+	// and is never evaluated at any time.
+	// When specified it needs to be a positive integer.
+	//
+	// +optional
+	SucceededCount *int
+}
+...
+
+// These are valid conditions of a job.
+const (
+	// JobSuccessCriteriaMet means the job has been succeeded.
+	JobSucceessCriteriaMet JobConditionType = "SuccessCriteriaMet"
+	...
+)
+```
+
+Moreover, we validate the following constraints for the `criteria`:
+- whether each criterion have at least one of the `succeededIndexes` or `succeededCount` specified.
+- whether the specified indexes in the `succeededIndexes` and
+  the number of indexes in the `succeededCount` don't exceed the value of `completions`.
+- whether `Indexed` is specified in the `completionMode` field.
+- whether the size of `succeededIndexes` is under 64Ki.
+- whether the `succeededIndexes` field has a valid format.
+- whether the `succeededCount` field has an absolute number.
+- whether the criteria haven't changed.
+- whether the successPolicies meet the `succeededCount <= |succeededIndexes|`, 
+where `|succeededIndexes|` means the number of indexes in the `succeededIndexes`. 
+
+
+### Evaluation
+
+Every time the pod condition are updated, the job-controller evaluates the successPolicies following the rules in order:
+
+- `succeededIndexes`: the job-controller evaluates `.status.completedIndexes` to see if a set of indexes is there.
+- `succeededCount`: the job-controller evaluates `.status.succeeded` to see if the value is `succeededCount` or more.
+
+After that, the job-controller adds a `SuccessCriteriaMet` condition instead of a `Failed` condition to `.status.conditions`
+and the job-controller terminates the lingering pods. At that time, `JobSuccessPolicy` is set to the `status.reason` field.
+
+Note that when the job meets one of successPolicies, other successPolicies are ignored.
+
+Finally, once all pods have terminated, the job-controller adds a `Complete` condition to `.status.conditions`.
+
+If any successPolicy isn't set, the job-controller adds both a `SuccessCriteriaMet` and a `Complete` conditions to the Job
+after the job finished.
+
+### Transition of "status.conditions"
+
+When the job with successPolicies is submitted, the job `status.conditions` transits in the following:
+Note that the Job doesn't have an actual `Running` condition in the `status.conditions`.
+
+```mermaid
+stateDiagram-v2
+    [*] --> Running
+    Running --> Failed: Exceeded backoffLimit
+    Running --> FailureTarget: Matched FailurePolicy with action=FailJob
+    FailureTarget --> Failed: Exceeded backoffLimit
+    Failed --> [*]
+    Running --> SuccessCriteriaMet: Matched SuccessPolicy
+    SuccessCriteriaMet --> Complete: All pods were terminated by successPolicies
```

Comment:
```suggestion
    SuccessCriteriaMet --> Complete: All pods are terminated
```
nit: this sounds informal, maybe not needed. The pods are actually terminated by the Job controller, in reaction the success policy.

Reply:
That makes sense.

Reply:
Done.